### ISSUE TITLE
[FEATURE] Embedding Model Integration

### ISSUE THREAD

**Issue by sfuentescamara** [sfuentescamara/credit-risk-rag-system#5]
---
title: "[FEATURE] Embedding Model Integration"
labels: rag, ml, llm, priority:critical, type:feature
milestone: "Milestone 1: Foundation & Setup"
assignees: ""
---

## 📋 Feature Overview
Integrate text embedding model for document vectorization to enable semantic search in the RAG system.

## 🎯 Objectives
- [ ] Select and integrate embedding model (Sentence-Transformers selected)
- [ ] Implement embedding generation service
- [ ] Add caching layer for embeddings
- [ ] Implement batch processing for efficiency
- [ ] Add error handling and retry logic
- [ ] Performance benchmarking and optimization

### Core Functionality
Generate high-quality vector embeddings from text documents to enable semantic similarity search in the RAG pipeline.

### Technology Choices

**Selected: Sentence-Transformers** ✅

**Model:** `all-MiniLM-L6-v2`

**Rationale:**
- **Pros:** Free, runs locally, good performance, 384-dimensional embeddings, fast inference
- **Cons:** Requires GPU for optimal speed with large batches (acceptable with CPU for MVP)

**Alternative Model:** `all-mpnet-base-v2`
- Better quality (768 dimensions)
- Slower inference
- Can upgrade later if quality is insufficient

**Alternative Options Considered:**

**Option 2: OpenAI Embeddings**
- Model: `text-embedding-3-small` or `text-embedding-3-large`
- Pros: High quality, no local resources needed, maintained by OpenAI
- Cons: Paid API ($0.02/1M tokens for small, $0.13/1M for large), rate limits, latency, external dependency

**Option 3: Cohere Embeddings**
- Model: `embed-english-v3.0` or `embed-multilingual-v3.0`
- Pros: Good quality, multilingual support
- Cons: Paid API, external dependency

**Recommended:** Sentence-Transformers (`all-MiniLM-L6-v2`) for MVP, can add OpenAI as fallback later

### Dependencies
- Depends on: #1 (Project Structure & Development Environment Setup)
- Depends on: #3 (Vector Database Selection & Setup) - for testing integration
- Blocks: #6 (Text Chunking Strategy Implementation)
- Blocks: #7 (RAG Retrieval Engine)

## 📁 Files to Create/Modify
- [ ] `src/rag/embeddings.py` - Embedding generation service
- [ ] `src/rag/cache.py` - Embedding cache implementation
- [ ] `src/config/models.py` - Model configuration
- [ ] `src/utils/retry.py` - Retry logic utilities
- [ ] `requirements.txt` - Add sentence-transformers and dependencies
- [ ] `tests/unit/test_embeddings.py` - Unit tests for embedding generation
- [ ] `tests/integration/test_embeddings_cache.py` - Integration tests with cache
- [ ] `tests/performance/test_embedding_speed.py` - Performance benchmarks
- [ ] `.env.example` - Add embedding model configuration
- [ ] `docs/embeddings.md` - Documentation for embedding service

### Development
- [ ] Install sentence-transformers package (`sentence-transformers>=2.2.0`)
- [ ] Create EmbeddingGenerator class with:
  - [ ] Model loading and initialization
  - [ ] Single text embedding generation
  - [ ] Batch embedding generation
  - [ ] Model caching (load once, reuse)
  - [ ] Device management (CPU/GPU auto-detection)
- [ ] Implement caching layer:
  - [ ] In-memory cache (LRU)
  - [ ] Optional disk cache for persistence
  - [ ] Cache key generation (hash-based)
  - [ ] Cache statistics
- [ ] Add configuration management:
  - [ ] Model name selection
  - [ ] Batch size configuration
  - [ ] Cache size limits
  - [ ] Device selection (CPU/CUDA)
- [ ] Error handling and retry logic:
  - [ ] Model loading failures
  - [ ] OOM (Out of Memory) handling
  - [ ] Batch size auto-adjustment
  - [ ] Exponential backoff for retries
- [ ] Logging for all operations
- [ ] Performance monitoring (time, memory usage)

### Testing
- [ ] Unit tests for EmbeddingGenerator (≥ 80% coverage)
- [ ] Test different text inputs (short, long, special characters)
- [ ] Test batch processing with various batch sizes
- [ ] Test cache hit/miss scenarios
- [ ] Test error scenarios (invalid input, OOM)
- [ ] Performance benchmarks:
  - [ ] Single embedding latency
  - [ ] Batch throughput
  - [ ] Cache performance impact
  - [ ] Memory usage profiling
- [ ] Integration tests with vector database

### Security
- [ ] No sensitive data in embeddings metadata
- [ ] Input validation (max length, encoding)
- [ ] Resource limits (prevent OOM attacks)

### Documentation
- [ ] Add docstrings to all functions/classes
- [ ] Document model selection rationale
- [ ] Create usage examples
- [ ] Document performance characteristics
- [ ] Cache usage guidelines
- [ ] Troubleshooting guide

### DevOps
- [ ] Model files handling (download, cache location)
- [ ] Docker configuration for model storage
- [ ] Environment variable configuration
- [ ] Resource requirements documentation (CPU/GPU/Memory)

## 🧪 Testing Criteria
```python
# Test embedding generation
from src.rag.embeddings import EmbeddingGenerator

# Initialize
generator = EmbeddingGenerator(model_name="all-MiniLM-L6-v2")

# Test single embedding
text = "Basel III capital requirements for banks"
embedding = generator.encode(text)

# Verify embedding properties
assert len(embedding) == 384  # Model dimension
assert all(isinstance(x, float) for x in embedding)
assert embedding.dtype == 'float32'

# Test batch processing
texts = [
    "Credit risk assessment",
    "Financial statement analysis",
    "Regulatory compliance"
]
embeddings = generator.encode_batch(texts, batch_size=32)
assert len(embeddings) == 3
assert embeddings.shape == (3, 384)

# Test similarity
from sklearn.metrics.pairwise import cosine_similarity
text1 = "credit risk evaluation"
text2 = "credit risk assessment"
text3 = "weather forecast"

emb1 = generator.encode(text1)
emb2 = generator.encode(text2)
emb3 = generator.encode(text3)

sim_12 = cosine_similarity([emb1], [emb2])[0][0]
sim_13 = cosine_similarity([emb1], [emb3])[0][0]

# Similar texts should have higher similarity
assert sim_12 > sim_13
assert sim_12 > 0.7  # High similarity threshold

# Test cache
embedding_cached = generator.encode(text)  # Should hit cache
assert generator.cache_stats()['hits'] > 0
```

```bash
# Performance benchmark
python tests/performance/test_embedding_speed.py

# Expected output:
# Single embedding: < 50ms (CPU), < 10ms (GPU)
# Batch 100 texts: < 2s (CPU), < 500ms (GPU)
# Cache hit: < 1ms
# Memory usage: < 500MB
```

## 📊 Success Metrics
- Model loads successfully on first run
- Single embedding generation < 50ms (CPU)
- Batch processing ≥ 100 texts/second (CPU)
- Cache hit rate > 80% in typical usage
- Memory usage < 500MB for model
- Test coverage ≥ 80%
- No memory leaks in long-running tests
- Embeddings have expected dimensionality (384)
- Similar texts have cosine similarity > 0.7

## 🔗 Related Resources
- [Sentence-Transformers Documentation](https://www.sbert.net/)
- [all-MiniLM-L6-v2 Model Card](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2)
- [MTEB Leaderboard](https://huggingface.co/spaces/mteb/leaderboard) - Embedding model benchmarks
- [Embedding Best Practices](https://www.pinecone.io/learn/sentence-embeddings/)
- [Text Embedding Guide](https://platform.openai.com/docs/guides/embeddings)

### Decisions Made
- **Embedding Model:** Sentence-Transformers `all-MiniLM-L6-v2` - Good balance of speed and quality for MVP
- **Dimension:** 384 - Sufficient for semantic search, faster than 768-dim models
- **Normalization:** L2 normalized by default (for cosine similarity)
- **Caching Strategy:** LRU cache with configurable size limit

### Implementation Notes

**EmbeddingGenerator Class Structure:**
```python
from sentence_transformers import SentenceTransformer
import numpy as np
from functools import lru_cache
import hashlib

class EmbeddingGenerator:
    """Generate embeddings using Sentence-Transformers."""

    def __init__(
        self,
        model_name: str = "all-MiniLM-L6-v2",
        device: str = None,
        cache_size: int = 1000
    ):
        """
        Initialize embedding generator.

        Args:
            model_name: Name of the sentence-transformers model
            device: 'cpu', 'cuda', or None for auto-detection
            cache_size: Number of embeddings to cache
        """
        self.model_name = model_name
        self.device = device or self._auto_detect_device()
        self.model = SentenceTransformer(model_name, device=self.device)
        self._cache = {}
        self._cache_size = cache_size
        self._cache_hits = 0
        self._cache_misses = 0

    def encode(
        self,
        text: str,
        normalize: bool = True
    ) -> np.ndarray:
        """Generate embedding for single text."""
        # Check cache
        cache_key = self._get_cache_key(text)
        if cache_key in self._cache:
            self._cache_hits += 1
            return self._cache[cache_key]

        # Generate embedding
        self._cache_misses += 1
        embedding = self.model.encode(
            text,
            normalize_embeddings=normalize,
            show_progress_bar=False
        )

        # Update cache
        self._update_cache(cache_key, embedding)

        return embedding

    def encode_batch(
        self,
        texts: list[str],
        batch_size: int = 32,
        normalize: bool = True
    ) -> np.ndarray:
        """Generate embeddings for batch of texts."""
        embeddings = self.model.encode(
            texts,
            batch_size=batch_size,
            normalize_embeddings=normalize,
            show_progress_bar=True
        )
        return embeddings

    def _get_cache_key(self, text: str) -> str:
        """Generate cache key from text."""
        return hashlib.md5(text.encode()).hexdigest()

    def _update_cache(self, key: str, embedding: np.ndarray):
        """Update LRU cache."""
        if len(self._cache) >= self._cache_size:
            # Remove oldest item (simple FIFO, can improve with OrderedDict)
            self._cache.pop(next(iter(self._cache)))
        self._cache[key] = embedding

    def cache_stats(self) -> dict:
        """Return cache statistics."""
        total = self._cache_hits + self._cache_misses
        hit_rate = self._cache_hits / total if total > 0 else 0
        return {
            'hits': self._cache_hits,
            'misses': self._cache_misses,
            'hit_rate': hit_rate,
            'size': len(self._cache)
        }

    def _auto_detect_device(self) -> str:
        """Auto-detect best available device."""
        import torch
        return 'cuda' if torch.cuda.is_available() else 'cpu'
```

**Configuration (config/models.py):**
```python
from pydantic import BaseModel

class EmbeddingConfig(BaseModel):
    """Embedding model configuration."""
    model_name: str = "all-MiniLM-L6-v2"
    device: str | None = None  # Auto-detect
    batch_size: int = 32
    cache_size: int = 1000
    normalize: bool = True
    max_seq_length: int = 256  # Model's max sequence length
```

**Environment Variables (.env.example):**
```bash
# Embedding Model Configuration
EMBEDDING_MODEL=all-MiniLM-L6-v2
EMBEDDING_DEVICE=cpu  # cpu, cuda, or auto
EMBEDDING_BATCH_SIZE=32
EMBEDDING_CACHE_SIZE=1000
```

### Questions/Clarifications Needed
- [x] Which embedding model? **Sentence-Transformers all-MiniLM-L6-v2 selected**
- [ ] GPU availability in production? (Currently supporting both CPU/GPU)
- [ ] Cache persistence required? (Currently in-memory only)
- [ ] Multi-language support needed? (Current model is English-optimized)
- [ ] Model update strategy? (To be defined)

### Performance Optimization Ideas
1. **Batch processing** - Always process in batches when possible
2. **Model quantization** - Can reduce model size by 4x with minimal quality loss
3. **ONNX conversion** - Faster inference with ONNX runtime
4. **GPU utilization** - Automatic GPU detection and usage
5. **Async processing** - For API endpoints (future enhancement)

## 📸 Screenshots/Diagrams

**Embedding Generation Flow:**
```
┌─────────────────────────────────────────────────────────────┐
│                    Text Input                                │
│         "Basel III capital requirements"                     │
└────────────────────────┬────────────────────────────────────┘
                         │
                         ▼
              ┌──────────────────────┐
              │  Cache Check         │
              │  (MD5 hash key)      │
              └─────┬───────┬────────┘
                    │       │
         Cache Hit  │       │  Cache Miss
                    │       │
                    ▼       ▼
         ┌──────────┐   ┌──────────────────────┐
         │ Return   │   │ Sentence-Transformers │
         │ Cached   │   │ Model (MiniLM)       │
         │ Embedding│   │                      │
         └──────────┘   │ - Tokenize           │
                        │ - Encode             │
                        │ - Pool               │
                        │ - Normalize          │
                        └──────┬───────────────┘
                               │
                               ▼
                    ┌──────────────────────┐
                    │  Update Cache        │
                    │  (LRU eviction)      │
                    └──────────┬───────────┘
                               │
                               ▼
                    ┌──────────────────────┐
                    │  Return Embedding    │
                    │  [0.1, 0.2, ..., ]   │
                    │  Shape: (384,)       │
                    └──────────────────────┘
```

**Batch Processing Flow:**
```
┌─────────────────────────────────────┐
│   List of Texts (N documents)       │
│   ["text1", "text2", ..., "textN"]  │
└──────────────┬──────────────────────┘
               │
               ▼
    ┌──────────────────────┐
    │ Split into Batches   │
    │ (batch_size=32)      │
    └──────┬───────────────┘
           │
           ▼
    ┌──────────────────────┐
    │ Process Each Batch   │
    │ in Parallel (GPU)    │
    └──────┬───────────────┘
           │
           ▼
    ┌──────────────────────┐
    │ Concatenate Results  │
    │ Shape: (N, 384)      │
    └──────────────────────┘
```

---

**Priority:** Critical
**Milestone:** Milestone 1: Foundation & Setup
**Sprint:** Sprint 1
